README.md
# nanako

[![npm](https://img.shields.io/npm/v/nanako.svg?style=flat-square)](https://www.npmjs.com/package/nanako)
![react](https://img.shields.io/npm/dependency-version/nanako/react?style=flat-square)
![next](https://img.shields.io/npm/dependency-version/nanako/next?style=flat-square)
![license](https://img.shields.io/npm/l/nanako.svg?style=flat-square)
![size](https://img.shields.io/github/repo-size/yinyanfr/nanako?style=flat-square)

Yet another static site generator.

## :star2: Features

- Designed for light novels
- CLI tools
- Sync tool (TODO)

## :green_book: Quick start

```bash
npx nanako create my-awesome-blog
```

## :wrench: CLI

```bash
Usage: nanako [options] [command]

Commands:
  add, a     Add a book or a chapter, or both.
  create, c  Create a nanako project.
  help       Display help
  version    Display version

Options:
  -h, --help     Output usage information
  -v, --version  Output the version number
  -y, --yes      Optional: skip the prompts and init the book or the chapter with the default meta.json.

Examples:
  - Create a nanako instance in folder my-awesome-blog.
  $ npx nanako create my-awesome-blog

  - Create a book.
  $ npx nanako add my-awesome-book

  - Create a chapter under my-awesome-book
  $ npx nanako add my-awesome-book my-awesome-chapter
```

## :books: Manual Content Management

### Putting your docs

```
docs
└─ lixia  // book
   ├─ chapter2  // chapter
   │  ├─ 1.md
   │  └─ meta.json
   ├─ intro
   │  ├─ archives
   │  │  ├─ 立夏序章简.epub
   │  │  └─ 立夏序章繁.epub
   │  ├─ images
   │  │  ├─ alice.jpg
   │  │  ├─ nanzhi1.jpg
   │  │  ├─ nanzhi2.jpg
   │  │  ├─ yuyao1.jpg
   │  │  ├─ yuyao2.jpg
   │  │  └─ yuyao3.jpg
   │  ├─ 1.md
   │  ├─ 10.md
   │  ├─ 11.md
   │  ├─ 12.md
   │  ├─ 13.md
   │  ├─ 14.尾声.md
   │  ├─ 15.后记.md
   │  ├─ 2.md
   │  ├─ 3.md
   │  ├─ 4.md
   │  ├─ 5.md
   │  ├─ 6.md
   │  ├─ 7.md
   │  ├─ 8.md
   │  ├─ 9.md
   │  └─ meta.json
   └─ meta.json
```

### Creating `meta.json` for books and chapters

```json
{
  "title": "立夏",
  "lang": "zh-Hans",
  "index": 1,
  "category": "novel"
}
```

### Using images in markdown

```markdown
![nanzhi1](./images/nanzhi1.jpg)
```

Will automatically redirect to `/images/nanzhi1.jpg`

### Modify `src/nanako.json`

```json
{
  "port": 20200,
  "footer": "对于本网站及展示的作品，作者保留一切权力。",
  "title": "小喵新姬的博客",
  "headTitle": "咕噜咕噜喵儿",
  "cookieConsent": true,
  "categories": {
    "novel": "小说",
    "casual": "随笔",
    "tutorial": "教程",
    "uncategorized": "未分类"
  },
  "cacheTTL": 600, // cache ttl for chapters fetch
  "cacheTTLLong": 3600 // cache ttl for books and sections fetch
}
```

## :rocket: Deployment

```bash
yarn build

yarn start

# or

NODE_ENV=production node nanako.js
```

Please refer to the [Next.js documentations](https://nextjs.org/docs) for anything that goes wrong with deployment.


CHANGELOG.md
# Change Log

## 1.2.0

`2023-04-17`

### Features

- [CLI] Added CLI
  - Create a nanako instance
  - add books / chapters

### Misc

- upgraded to antd 5 and next 13


src/cli/commands.js
#!/usr/bin/env node

const fs = require("node:fs");
const path = require("node:path");
const ora = require("ora");
const prompts = require("prompts");
const { isNanakoRoot } = require("./lib");

// src/cli/command.js
const rootPath = path.join(__dirname, "..", "..");

const createCommand = async (name, sub, options) => {
  const target = sub?.[0] ?? ".";
  const createPath = path.resolve(target);

  const configs = {
    categories: {
      novel: "novel",
      casual: "casual",
      tutorial: "tutorial",
      uncategorized: "uncategorized",
    },
    cacheTTL: 600,
    cacheTTLLong: 3600,
  };
  const questions = [
    {
      type: "text",
      name: "title",
      message: "What is the title of your site?",
    },
    {
      type: "text",
      name: "headTitle",
      message: "What is the browser tab title of your site?",
      initial: (prev) => prev,
    },
    {
      type: "text",
      name: "footer",
      message: "What is the footer text of your site?",
    },
    {
      type: "number",
      name: "port",
      message: "Which port will your site be running on?",
    },
  ];
  if (!options?.yes) {
    try {
      const response = await prompts(questions);
      Object.keys(response).forEach((key) => {
        configs[key] = response[key];
      });
    } catch (error) {
      throw error;
    }
  }

  ["public", "src"].forEach((folder) => {
    const spinner = ora(`Copying ${folder}`).start();
    fs.cpSync(path.join(rootPath, folder), path.join(createPath, folder), {
      recursive: true,
    });
    spinner.succeed(`Success: Copying ${folder}`);
  });
  [
    ".eslintrc.json",
    "LICENSE",
    "nanako.js",
    "next-env.d.ts",
    "next.config.js",
    "package.json",
    "README.md",
    "tsconfig.json",
  ].forEach((file) => {
    const spinner = ora(`Copying ${file}`).start();
    fs.cpSync(path.join(rootPath, file), path.join(createPath, file));
    spinner.succeed(`Success: Copying ${file}`);
  });

  if (!options.yes) {
    const spinner = ora(`Writing configs`).start();
    const configPath = path.join(createPath, "src", "nanako.json");
    fs.writeFileSync(configPath, JSON.stringify(configs, null, 2));
    spinner.succeed(`Writing configs`);
  }

  console.log(`\nYour nanako project is created at ${path.resolve(target)}\n`);
  console.log("You can now run:\n");
  console.log(`\tcd ${target}`);
  console.log(`\tnpm i # or yarn`);
  console.log(`\tnpx . help`);
};

const addCommand = async (name, sub, options) => {
  const [book, chapter] = sub ?? [];
  if (!isNanakoRoot()) {
    throw new Error(
      "Please run this command at the root folder of your Nanako instance."
    );
  }
  if (!book) {
    throw new Error("Please specify the name of the book.");
  }

  const bookPath = path.resolve("docs", book);
  if (!chapter) {
    const initials = {
      title: book,
      category: "novel",
      lang: "zh-Hans",
      index: 1,
    };
    const questions = [
      {
        type: "text",
        name: "title",
        message: "What is the title of your book?",
        initial: initials.title,
      },
      {
        type: "text",
        name: "category",
        message: "What is the category of your book?",
        initial: initials.category,
      },
      {
        type: "text",
        name: "lang",
        message: "Which language (locale) is your book?",
        initial: initials.lang,
      },
      {
        type: "number",
        name: "index",
        message: "What is the index (order) of your book?",
        initial: initials.index,
      },
    ];
    const meta = options?.yes ? initials : await prompts(questions);
    fs.mkdirSync(bookPath, { recursive: true });
    fs.writeFileSync(
      path.join(bookPath, "meta.json"),
      JSON.stringify(meta, null, 2)
    );
    console.log(`Success: Your book is created at ${bookPath}`);
  } else {
    if (!fs.existsSync(bookPath)) {
      console.warn(
        `Please create the book first by using:\nnpx nanako add ${book}`
      );
      return 1;
    }
    const chapterPath = path.join(bookPath, chapter);
    const initials = {
      title: chapter,
      index: 1,
    };
    const questions = [
      {
        type: "text",
        name: "title",
        message: "What is the title of your chapter?",
        initial: initials.title,
      },
      {
        type: "number",
        name: "index",
        message: "What is the index (order) of your chapter?",
        initial: initials.index,
      },
    ];
    const meta = options?.yes ? initials : await prompts(questions);
    fs.mkdirSync(chapterPath, { recursive: true });
    fs.writeFileSync(
      path.join(chapterPath, "meta.json"),
      JSON.stringify(meta, null, 2)
    );
    console.log(`Success: Your chapter is created at ${chapterPath}`);
  }
};

module.exports = {
  createCommand,
  addCommand,
};
